CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+detect or protect against errors.
+
+The class contract represents low-level software requirements, which the methods
+of the class are traced to by being proximal to the source code.  In addition,
+the methods of the derived class are traced to the requirements of the base
+class (along with any requirements added by the derived class).  The derived
+class must meet the requirements of the base class in order to maintain type
+consistency.
+
+Type consistency can be easier to verify when interface and implementation
+inheritance are not combined.  Verification of interface inheritance is
+straightforward with abstract base classes that contain only pure-virtual
+function declarations and no member variables.  [^Sutter36]  Because the
+abstract base class defines only interface requirements, there is no way for
+derived classes to put the base class into an undesired state.  See
+[this example](guidelines_inheritance_examples.md#AbstractBaseClass) for more
+information.
+
+When mixing interface and implementation inheritance is necessary, consider
+using the non-virtual interface idiom (NVI), which is, in summary:
+
+- Prefer to make interfaces non-virtual, using the Template Method pattern.
+- Prefer to make virtual functions private.
+- Only if derived classes need to invoke the base implementation of a virtual
+  function, make the virtual function protected.
+- A base class destructor should be either public and virtual, or protected
+  and non-virtual.
+
+Dispatching calls through the non-virtual interface of the base class
+to the virtual methods of the derived classes provides a single place to
+specify and/or check preconditions and postconditions of the virtual methods.
+It also moves the dynamic dispatch from the calling routine into the base
+class method.  This allows the base class and its derived class hierarchy
+to be tested as a self-contained unit.
+
+A base class destructor is a special case, because it is undefined behavior to
+`delete` a derived class object using a pointer to its base class, unless the
+base class destructor is virtual.  To avoid the undefined behavior, the base
+class destructor should be either public and virtual (to call the appropriate
+derived class constructor upon deletion), or protected and non-virtual (to
+disallow deletion of a derived class).
+
+Another alternative to virtual interface inheritance and runtime substitution
+is the Curiously Recurring Template Parameter (CRTP) idiom.
+The CRTP allows a non-virtual interface to be defined that is implemented by a
+class that is substituted at compile-time, resulting in static dispatch and
+opportunities to catch errors during compilation.
+
+The `override` keyword should be used when overriding virtual methods.  `final`
+should be used for leaf classes in a class hierarchy.  Class methods should be
+marked ether `final` or `override`, but not both.
+
+Virtual methods should not be called inside a constructor or destructor.
+
+**Multiple Inheritance**
```

Comment:
I am not the biggest fan of multiple inheritance.

Reply:
I'm fine with the way it is outlined here. Single class, multiple interface inheritance is built-in to basically every language that isn't C++.